### PR TITLE
[FLINK-25156][table-planner] Support DISTINCT in CastRules

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Signed-off-by: slinkydeveloper <francescoguard@gmail.com>

## What is the purpose of the change

Support `DISTINCT` type in `CastRule`s

## Brief change log

* Now `CastRuleProvider` unwraps the `DISTINCT` types before resolving the cast rule.

## Verifying this change

Added `CastRuleProviderTest` to test the introduced unwrapping behaviour. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRuleProvider.java [hunk 1]
```diff
@@ -213,4 +217,11 @@ private CastRuleProvider addRule(CastRule<?, ?> rule) {
                 .findFirst()
                 .orElse(null);
     }
+
+    private LogicalType unwrapDistinct(LogicalType logicalType) {
+        if (logicalType.is(LogicalTypeRoot.DISTINCT_TYPE)) {
```
Comment (matriv) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRuleProvider.java:222]: This is just a suggestion, and if we agree it can also be done as another PR: Maybe we can use the `accept()` with a visitor that extracts the underlying type directly, or return the same type (for the simple types) and can be used for other types, like `ARRAY` & `RAW`.

**Review by matriv — APPROVED** [review]
LGTM!

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRuleProvider.java (hunk 1 above)
Comment (matriv) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRuleProvider.java:222]: Hm, on second thought this must be done before actually resolving the type, so if it's an `ARRAY` then we don't want to extract the underlying type.

Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRuleProvider.java:222]: I think it's fine to don't deeply unwrap it, for example assume you wanna cast using `ARRAY<DISTINCT<INT>>` to `ARRAY<LONG>`. Because in `ArrayToArrayCastRule` the method `CastRuleProvider#exists` and `CastRuleProvider#resolve` are invoked for the array item type, the `DISTINCT<INT>` will be correctly unwrapped to `INT`. Let me add a test just to make sure this works

Comment (matriv) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRuleProvider.java:222]: Thx for adding this test!

**Review by twalthr — APPROVED** [review]
LGTM if build is green
